File: employee-service/Middleware/CorrelationIdMiddleware.cs
namespace EmployeeService.Middleware;

public class CorrelationIdMiddleware
{
    private const string HeaderName = "X-Correlation-Id";
    private readonly RequestDelegate _next;

    public CorrelationIdMiddleware(RequestDelegate next) => _next = next;

    public async Task InvokeAsync(HttpContext context)
    {
        var cid = context.Request.Headers[HeaderName].FirstOrDefault();
        if (string.IsNullOrWhiteSpace(cid))
            cid = Guid.NewGuid().ToString();

        context.Items[HeaderName] = cid;
        context.Response.Headers[HeaderName] = cid;

        await _next(context);
    }
}

File: auth-service/src/correlationId.middleware.js
import { v4 as uuidv4 } from 'uuid';

export function correlationIdMiddleware(req, res, next) {
  let correlationId = req.headers['x-correlation-id'];
  if (!correlationId) {
    correlationId = uuidv4();
    req.headers['x-correlation-id'] = correlationId;
  }
  res.setHeader('x-correlation-id', correlationId);
  req.correlationId = correlationId;
  next();
}